### ISSUE TITLE
Circular references

### ISSUE THREAD

**Issue by slinkydeveloper** [RepreZen/KaiZen-OpenApi-Parser#64]
It seems that loading an api spec with schemas with circular references causes the `parse()` method throws `StackOverflowException`.

### Steps to reproduce
Try to load this spec:
```yaml
openapi: "3.0.0"
info:
  version: 1.0.0
  title: bundle schema test
paths:
  /pets:
    get:
      operationId: findPets
      responses:
        200:
          description: "hello"
components:
  schemas:
    schema1:
      title: schema1
      type: object
      properties:
        a:
          $ref: '#/components/schemas/schema2'
        b:
          type: string
        c:
          type: number
    schema2:
      title: schema2
      type: object
      properties:
        a:
          type: array
          items:
            $ref: '#/components/schemas/schema1'
```

## Some investigation
I've done some investigation to try to fix that bug. It can be resolved processing the references nodes during the spec walk.
For example: in the spec above when you call `parse()` the parser **doesn't have** to resolve the `schema1.properties.a` reference, because it goes on `StackOverflowException`. If you call `model....get("a")` It has to solve the reference and give me a `SchemaImpl` object. Then, the next time you call `model....get("a")`, the reference is already solved.
Of course it decreases a little bit the perfomances but I think it can be enabled/disabled with a flag during parser initialization

## I tried to fix it :smiley:
I tried to fix it, but I haven't understand some code. Sorry for that, I attach my progresses: [circular-reference.zip](https://github.com/RepreZen/KaiZen-OpenApi-Parser/files/1219738/circular-reference.zip) (this patch is referenced to [branch master of my fork](https://github.com/slinkydeveloper/KaiZen-OpenApi-Parser)). Inspecting the code, it seems that after trying to walk into a `JsonOverlay` of a reference, it has the right json deferenced but it doesn't have values (in the `BundleSchemaTest` you can look that `model.getSchemas().get("schema2").getProperties().get("a").getTitle()` returns null, try to inspect it). Also I think that my attempt to fix it fails on `ObjectOverlay#fromJson()` function.

Also in this code there's a function inside `SchemaImpl` called `bundleSchema()`. This is a small draft attempting to replicate [this functionality](https://github.com/BigstickCarpet/json-schema-ref-parser/blob/master/docs/ref-parser.md#bundleschema-options-callback) (I need the schema to validate requests for my [vertx-web project](https://github.com/vert-x3/vertx-web/pull/666)). Of course I can't test it because parser doesn't load the project

**Comment by andylowry** [comment]
@slinkydeveloper This is now addressed, and references are now fully visible in the API. The API is documented in a new markdown document called `API-Overview`.

**Comment by andylowry** [comment]
Addressed in #94
